### PR TITLE
Garanti Circuits

### PR DESCRIPTION
**Pull request by richardliang**
### Circuits

- [x] Regexes
   - [x] Garanti Payer Details
   - [x] Garanti Payment Details
   - [x] Garanti Subject
   - [x] Garanti Timestamp
- [x] Garanti Send
- [x] Garanti Registration
- [x] Body suffix hasher
- [x] Body hash v2 that support simple canonicallization

### Tasks
- [x] Move the common regexes to a separate PR and use them in the Paylah circuits (#248)
- [x] Verify body hash extraction is needed for both circuits. Move body hash regex to circuit 1, and allow as user input to circuit 2. Verify bh1 = bh2 in contract
- [x] Optimize multi_or in body hash regex v2
- [x] Finalize constants for Garanti
- [x] Ensure reply-to doesn't pass for to regex
- [x] Compile circuits with complete email length
- [x] Ensure nullifier tests pass
- [x] Add CI
- [x] Review regexes for optimization
- [x] Review fixes

### Build
- [x] Clean compile (remove logs)
- [x] Gen keys
- [x] Gen verifier
- [x] Gen calldata
- [ ] Upload
- [x] Publish package

Send calldata
["0x0dc8f2891b315ac064ef74c169d0d8ac44f63c39eccf0170093e8485ed9e92bb", "0x151e64dd9470390e5eb7e94452dc4078c04cb9e476902fb2a7c69854d294345d"],[["0x105ecb7d06532f5e4e9e82ff16d9dc853430d964b237c359222167f4f86d41a6", "0x08ddfcc2d8c29702bc7952a65a4e2e0239daa18bb299ac3069aaf2ee8fb4d0d3"],["0x263ebeac2d88b084cb87983504db305ec811489db7c67aaca73188756cc42de7", "0x1b2f21517a257805ed0c1179080b6ae8fdfd2f997dc47054efcb7e2d75548a94"]],["0x1c46a8acadeba9235863bb7e9145661c516988c39fe2b3de08fcde2e271746ad", "0x110a6249e23c8fe793cc6e8b63e4e7d3aab1d5c276f9d5e8a17b069944047071"],["0x03a9c8babd6b4ad94d711f3ffbee84b7aa69f4cb0dd08d491c5a5c32eca15f60","0x000000000000000000000000000000000de437f1a84c19f842cf57e8050277ee","0x0000000000000000000000000000000085fb4e44aa54c94873414ecbf4487122","0x00000000000000000000000000000000ec2c7e8ebdd0cf3ecdbd2a47c25f6882","0x000000000000000000000000000000007d1c9c7f962023b3b2ce05fc22905b9b","0x0000000000000000000000000000000000000000000000000069746e61726167","0x00000000000000000000000000000000000000000000000000672e6f666e6940","0x000000000000000000000000000000000000000000000000006269746e617261","0x000000000000000000000000000000000000000000000000006d6f632e617662","0x000000000000000000000000000000000000000000000000000000000072742e","0x0000000000000000000000000000000000000000000000000036313433303731","0x0000000000000000000000000000000000000000000000000000000000313638","0x0000000000000000000000000000000000000000000000000030302031345254","0x0000000000000000000000000000000000000000000000000031313030203130","0x0000000000000000000000000000000000000000000000000030203437353620","0x0000000000000000000000000000000000000000000000000035333720373437","0x0000000000000000000000000000000000000000000000000000000035302030","0x000000000000000000000000000000000000000000000000000030302c303038","0x0000000000000000000000000000000000000000000000000000000000000000","0x071caee57e1bb6d2542fffa8b70fdd31f73ce937e8aa12735701d2481b7b0228","0x1234815a8ff6956b9b13bbc40073354659eb0ed3e1c1c271be7315d25a3e9105","0x0000000000000000000000000000000000000000000000000000000000003039"]

Registration calldata
["0x0389d132b5d2e1d9df89b8ebd21e6116b4f2640b4a191d79292bd3c4fcf8b16a", "0x19987648fefb4e66335c9f045713a0434fc4ba8e6b380767207129d744b8ffc2"],[["0x0a2ee9c1de104203628d6dc3224f7630ee3df8b22d4d5b99659cdfa6b3507fd1", "0x2af41646939b18c265f46b101aa964df395273c5269d63e288eb70893bde60d2"],["0x0aa16b290262094f120a76fc6d53490e373ccc1325b0c9d98573d19f78b54eab", "0x2499ae14df27902e06077a9dd3a2bb8f0c492bb2d67e2f0f6f21e611cf3517b0"]],["0x010001222e45ad8f4dadcdb67608a4bc6462aac6436a6b52372f3fa445436429", "0x1b067ff66aac409e7c976ad48738686f2fd0fd817abe02999cc5594282469c5f"],["0x03a9c8babd6b4ad94d711f3ffbee84b7aa69f4cb0dd08d491c5a5c32eca15f60","0x000000000000000000000000000000000de437f1a84c19f842cf57e8050277ee","0x0000000000000000000000000000000085fb4e44aa54c94873414ecbf4487122","0x00000000000000000000000000000000ec2c7e8ebdd0cf3ecdbd2a47c25f6882","0x000000000000000000000000000000007d1c9c7f962023b3b2ce05fc22905b9b","0x0000000000000000000000000000000000000000000000000069746e61726167","0x00000000000000000000000000000000000000000000000000672e6f666e6940","0x000000000000000000000000000000000000000000000000006269746e617261","0x000000000000000000000000000000000000000000000000006d6f632e617662","0x000000000000000000000000000000000000000000000000000000000072742e","0x071caee57e1bb6d2542fffa8b70fdd31f73ce937e8aa12735701d2481b7b0228"]

Body suffix hasher calldata (same for registration and send)
["0x2e4e61b91921f2598d1d62bdcc0d6db92176733fffbff0278d6f4b8f74410129", "0x1ec5e75f6fbe487a3c22e833b19748fe395659b9310da5c771339121f60a9836"],[["0x23a63c99ce250a9c24e9b7e39401e2b1f2649c6868508028d9a2c1c2c8ab6232", "0x2a3c9a61a4345a6ba24b3bbe99debec89ca99542879ab9f2ecd39279f2664fb2"],["0x164de92e0add3a56fd779022d2606f088e21c302bbdd18bd72d57fc193f4038a", "0x1521f487a3708e7e16649d398191b9e378cd5f18b593600b44ce87cd7dc69210"]],["0x2e66d7f1c361ffc8b1201ae904321acc95717435ae2aff8f1822090496faee46", "0x3021feabfabed7c446ba0aec47108b6d90204663e457b1cb4487d41ec5d7e7a6"],["0x000000000000000000000000000000000de437f1a84c19f842cf57e8050277ee","0x0000000000000000000000000000000085fb4e44aa54c94873414ecbf4487122","0x00000000000000000000000000000000ec2c7e8ebdd0cf3ecdbd2a47c25f6882","0x000000000000000000000000000000007d1c9c7f962023b3b2ce05fc22905b9b"]

### CODE REVIEW

File: circuits-circom/circuits/common-v2/regexes/body_hash_regex_v2.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
+    "parts": [
+        {
+            "is_public": false,
+            "regex_def": "((\r\n)|^)(d|D)(k|K)(i|I)(m|M)-(s|S)ignature:( )?"
+        },
+        {
+            "is_public": false,
+            "regex_def": "((a|b|c|d|e|f|g|h|i|j|k|l|m|n|o|p|q|r|s|t|u|v|w|x|y|z)+=[^;]+;(\r|\n|\u0009| )+)+bh="
```
Comment (0xSachinK) [inline circuits-circom/circuits/common-v2/regexes/body_hash_regex_v2.json:9]: Should this rather be `(\r\n\u0009| )?`
I.e. Either a `\r\n\u0009` or `space`. And only one of any of them?

File: circuits-circom/circuits/garanti/garanti_send.circom [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+
+// Args:
+// * max_header_bytes = 1024 is the max number of bytes in the header
+// * max_body_bytes = 2496 is the max number of bytes in the body after precomputed slice
+// * n = 121 is the number of bits in each chunk of the modulus (RSA parameter)
+// * k = 17 is the number of chunks in the modulus (RSA parameter)
+// * pack_size = 7 is the number of bytes that can fit into a 255ish bit signal (can increase later)
+component main { public [ intent_hash ] } = GarantiSendEmail(1024, 2496, 121, 17, 7);
```
Comment (0xSachinK) [inline circuits-circom/circuits/garanti/garanti_send.circom:234]: Max header bytes can be 512.

**Review by 0xSachinK — CHANGES_REQUESTED** [review]
1. `DividedHasher` circuit doesn't need to assert equality. That can be done by the contract.
2. Let's move the `DividedHasher_10752` to garanti folder. Common folders only host templates.
3. The packing logic for intermediate hash and body hash is scattered across multiple files. Can instead make that a util and reuse it.
4. Instead of implementing custom `EmailVerifierHeader` we can just skip the body hash check in zk-email's `EmailVerifier` and then implement body hash logic in our custom circuit. Keeping the abstraction clean and audited and non-audited parts separate.

File: circuits-circom/circuits/common-v2/regexes/body_hash_regex_v2.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (richardliang) [reply circuits-circom/circuits/common-v2/regexes/body_hash_regex_v2.json:9]: `\t` isnt a supported character in zk-regex apparently

[the remark this answers is not in the document]
Comment (richardliang) [reply circuits-circom/circuits/common-v2/regexes/body_hash_regex_v2.json:9]: i believe for `simple`, space is also allowed in addition to tabs. Could be fine with this regex

**Pull request merged**